### DISCUSSION TITLE
zbbridge - SENSOR messages not published on teleperiod or status 8

### DISCUSSION THREAD

**Discussion by trichromic-ab** [arendst/Tasmota#14545] [Q&A] [answered]
### PROBLEM DESCRIPTION

This is my first posting, so hopefully all done correctly. I'm fairly new to Tasmota but I've been struggling to understand this issue for a few days and now suspect it might be a bug.

Using a Sonoff Zigbee Bridge with v10.1.0 of tasmota-zbbridge.bin, it seems that the SENSOR messages are not being published to MQTT on the teleperiod cycle. Only the STATE messages are published. I'm getting no SENSOR information.
I am having to wait for my Zigbee temperature and humidity sensors to change value before a SENSOR message comes through
Sometimes that can take a very long time, like over an hour. So Home Assistant is not getting any initial measurements.
I tried using STATUS 8 & STATUS 10 commands, as suggested in the notes about the SonoffPow but that does not trigger a publishing of the SENSOR messages.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): _____ Sonoff ZB Bridge
- [x] Tasmota binary firmware version number used: _____ 10.1.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [ ] Flashing tools used: _____
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
01:38:34.731 MQT: stat/tasmota-zb2/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
01:38:34.943 MQT: stat/tasmota-zb2/RESULT = {"Module":{"75":"Sonoff ZbBridge"}}
01:38:35.157 MQT: stat/tasmota-zb2/RESULT = {"GPIO0":{"320":"Led_i1"},"GPIO1":{"3552":"Zigbee Tx"},"GPIO2":{"0":"None"},"GPIO3":{"3584":"Zigbee Rx"},"GPIO4":{"5312":"Zigbee Rst1"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"640":"I2C SDA"},"GPIO13":{"576":"LedLink_i"},"GPIO14":{"608":"I2C SCL"},"GPIO15":{"0":"None"},"GPIO16":{"32":"Button1"},"GPIO17":{"0":"None"}}
01:38:55.399 MQT: tele/tasmota-zb2/STATE = {"Time":"2022-01-21T01:38:55","Uptime":"0T02:27:39","UptimeSec":8859,"Vcc":3.538,"Heap":27,"SleepMode":"Normal","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"SSID2","BSSId":"02:C7:BF:14:9F:24","Channel":6,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}
01:39:25.395 MQT: tele/tasmota-zb2/STATE = {"Time":"2022-01-21T01:39:25","Uptime":"0T02:28:09","UptimeSec":8889,"Vcc":3.538,"Heap":27,"SleepMode":"Normal","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"SSID2","BSSId":"02:C7:BF:14:9F:24","Channel":6,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
01:41:31.463 MQT: stat/tasmota-zb2/STATUS = {"Status":{"Module":75,"DeviceName":"tasmota-zb2","FriendlyName":["tasmota-zb2"],"Topic":"tasmota-zb2","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
01:41:31.470 MQT: stat/tasmota-zb2/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota-zbbridge.bin.gz","RestartReason":"Software/System restart","Uptime":"0T02:30:15","StartupUTC":"2022-01-20T22:11:16","Sleep":50,"CfgHolder":4617,"BootCount":33,"BCResetTime":"2021-03-21T20:28:55","SaveCount":54,"SaveAddress":"1FB000"}}
01:41:31.474 MQT: stat/tasmota-zb2/STATUS2 = {"StatusFWR":{"Version":"10.1.0(zbbridge)","BuildDateTime":"2021-12-08T14:47:37","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":160,"Hardware":"ESP8266EX","CR":"477/699"}}
01:41:31.481 MQT: stat/tasmota-zb2/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["SSID2",""],"TelePeriod":30,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A00000000000000","00000480","00006002","00004000"]}}
01:41:31.490 MQT: stat/tasmota-zb2/STATUS4 = {"StatusMEM":{"ProgramSize":671,"Free":1120,"Heap":23,"ProgramFlashSize":2048,"FlashSize":2048,"FlashChipId":"1540A1","FlashFrequency":40,"FlashMode":3,"Features":["00000809","0F1007C6","04400001","00000003","00000000","00000000","00020080","00200000","00000000"],"Drivers":"1,2,4,7,9,10,12,20,23,38,41,50","Sensors":"1"}}
01:41:31.496 MQT: stat/tasmota-zb2/STATUS5 = {"StatusNET":{"Hostname":"tasmota-zb2","IPAddress":"192.168.1.32","Gateway":"192.168.1.254","Subnetmask":"255.255.255.0","DNSServer1":"208.67.220.220","DNSServer2":"208.67.222.222","Mac":"C8:2B:96:04:BC:82","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
01:41:31.499 MQT: stat/tasmota-zb2/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.1.19","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_04BC82","MqttUser":"homeassistant","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
01:41:31.504 MQT: stat/tasmota-zb2/STATUS7 = {"StatusTIM":{"UTC":"2022-01-21T00:41:31","Local":"2022-01-21T01:41:31","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"08:33","Sunset":"17:29"}}
01:41:31.507 MQT: stat/tasmota-zb2/STATUS10 = {"StatusSNS":{"Time":"2022-01-21T01:41:31"}}
01:41:31.513 MQT: stat/tasmota-zb2/STATUS11 = {"StatusSTS":{"Time":"2022-01-21T01:41:31","Uptime":"0T02:30:15","UptimeSec":9015,"Vcc":3.538,"Heap":23,"SleepMode":"Normal","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"SSID2","BSSId":"02:C7:BF:14:9F:24","Channel":6,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}}

```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

No more details provided in console

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Set teleperiod to 30 and watching console messages, no SENSOR messages are shown, only STATE messages every 30 seconds which don't have any sensor values included. Only by forcing a temperature change I can see any SENSOR message published.
The ZB Bridge does have measurements from all the sensors that are paired, as can be seen from the devices status on the main web page. It simply doesn't send any data periodically, only if the measurements change.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

With a Status 8 command I just get this result. I would expect sensor measurement data to be included here or a SENSOR message published:

01:47:45.316 MQT: stat/tasmota-zb2/STATUS8 = {"StatusSNS":{"Time":"2022-01-21T01:47:45"}}

This is all that comes on each teleperiod (currently 30 secs):

01:49:25.395 MQT: tele/tasmota-zb2/STATE = {"Time":"2022-01-21T01:49:25","Uptime":"0T02:38:09","UptimeSec":9489,"Vcc":3.538,"Heap":27,"SleepMode":"Normal","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"SSID2","BSSId":"02:C7:BF:14:9F:24","Channel":6,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:03"}}
01:49:55.431 MQT: tele/tasmota-zb2/STATE = {"Time":"2022-01-21T01:49:55","Uptime":"0T02:38:39","UptimeSec":9519,"Vcc":3.538,"Heap":27,"SleepMode":"Normal","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"SSID2","BSSId":"02:C7:BF:14:9F:24","Channel":6,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:03"}}
01:50:25.389 MQT: tele/tasmota-zb2/STATE = {"Time":"2022-01-21T01:50:25","Uptime":"0T02:39:09","UptimeSec":9549,"Vcc":3.540,"Heap":27,"SleepMode":"Normal","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"SSID2","BSSId":"02:C7:BF:14:9F:24","Channel":6,"Mode":"11n","RSSI":56,"Signal":-72,"LinkCount":1,"Downtime":"0T00:00:03"}}
 

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by ascillato** [comment, marked as answer]
Hi, the Sonoff ZigBee Bridge will publish ONLY the messages it gets and when it receive it. It won't store any sensor data from other zigbee devices.

**Reply by trichromic-ab** [reply]
Hi and thanks for reviewing my possible bug report and concluding its by design.

However I am confused because the web interface shows stored state value and these appear very soon after a bridge is restarted (as screenshot here). So the bridge does seem to be holding the last received values.

![image](https://user-images.githubusercontent.com/80370469/150523575-bc77cdfa-ff67-41ee-8cec-e12849397df9.png)

However those initial values don't get published to MQTT, either on restart or on the teleperiod cycle.
So any client has to wait for the next time a value changes in the sensor before it gets its first SENSOR message.
From experience, this can be a very long time, maybe more than an hour.

It doesn't even seem to be possible to request a refresh message using something like the Status 8 command.

I can't be sure as I've not been using Tasmota for long but I originally flashed with v9 and don't remember this being an issue then. I've only dug into this concern since I did a v10 update and found that Home Assistant is displaying 'no data' for a long time after a restart or a config change.

I realise that the ZigBee sensors do not remain sending messages at intervals in order to save battery power. So the bridge has to wait passively for a state change. But surely it makes sense for the bridge (being powered) to send what it knows about sensors at regular intervals and on the request of a client. Or am I missing the point on some design concept?

Regards,
Alex

**Reply by ascillato** [reply]
If the bridge sends the previous information that it has instead of what the sensor sends, your bridge will be lying to you because that information is old. The bridge only retransmits what it receives (because it is a ZigBee to Wi-Fi bridge). That is by ZigBee design.

Take for example the case which your Kitchen_sensor run out of battery. You will continue seeing an old value if the bridge continue sending to you old data and you will never know that aren't real values.

**Reply by trichromic-ab** [reply]
Hi,

I certainly can appreciate that scenario.
But if the stored values that are held by the bridge have a timestamp for when they were last updated, the client can choose whether they are valid or stale and thus use them or ignore them. It should not be up to the publishing device to control their availability to the client. 

However, this discussion about ZbInfo being missing from the documentation, has just answered part of my question:
[(https://github.com/arendst/Tasmota/discussions/14542#discussion-3816573)]

When I send a ZbInfo command to a bridge, I get this response:

```lua
13:31:21.657 MQT: tele/tasmota-zb2/SENSOR = {"ZbInfo":{"kitchen_sensor":{"Device":"0x7199","Name":"kitchen_sensor","IEEEAddr":"0x00124B0022686759","ModelId":"TH01","Manufacturer":"eWeLink","Endpoints":[1],"Config":["T01","O01"],"Temperature":16.85,"Humidity":43.95,"Power":0,"Reachable":true,"BatteryPercentage":83,"LastSeen":2392,"LastSeenEpoch":1642765889,"LinkQuality":68}}}
13:31:21.665 MQT: tele/tasmota-zb2/SENSOR = {"ZbInfo":{"dining_room_sensor":{"Device":"0x0169","Name":"dining_room_sensor","IEEEAddr":"0x00124B0022686D24","ModelId":"TH01","Manufacturer":"eWeLink","Endpoints":[1],"Config":["T01","O01"],"Temperature":15.87,"Humidity":46.64,"Power":0,"Reachable":true,"BatteryPercentage":58,"LastSeen":57,"LastSeenEpoch":1642768224,"LinkQuality":136}}}
13:31:21.668 MQT: stat/tasmota-zb2/RESULT = {"ZbInfo":"Done"}
```
That's just what I've been hoping for. I had expected to see this come with every teleperiod. 
Maybe I should try to add a service to HA so that I can refresh on demand.

**Reply by trichromic-ab** [reply]
PS. The documentation for TelePeriod does say that both STATE and SENSOR messages should be published:

![image](https://user-images.githubusercontent.com/80370469/150530873-e1a569cf-6247-44d1-909f-bfd423752ad9.png)

**Reply by ascillato** [reply]
> PS. The documentation for TelePeriod does say that both STATE and SENSOR messages should be published:

Yes, but as stated before, that does not apply to the ZB-Bridge. The ZB Bridge in itself does not have any local sensor. It retransmits what it receives (because it is a ZigBee to Wi-Fi bridge). The **_teleperiod sensors_** is for **local sensors** in the device itself not in other devices.

**Reply by ascillato** [reply]
> When I send a ZbInfo command to a bridge, I get this response:

Yes, you can use that or `ZbStatus3 <device>`

**Reply by barbudor** [reply]
If you use `SensorRetain 1` and `SetOption89 1`, each sensor message will be published on its own sub-topic and the message will be retained.
So you MQTT broker will have in memory the last value and any client connecting to it will retreive the last published value.
https://tasmota.github.io/docs/Zigbee/#reading-sensors

**Reply by trichromic-ab** [reply]
Thanks both. I will give those suggestions a try.

20 years ago I was a factory automation engineer working with industrial PLCs. I had the luxury of being able to determine any specific input state by polling cycles. Having to wait indeterminately for valid state with ZigBee is something I need to get used to.

**Reply by trichromic-ab** [reply]
> Yes, you can use that or `ZbStatus3 <device>`

Thanks. I tested and note that ZbStatus3 <device> gives a RESULT message response, whereas a ZbInfo gives a SENSOR message response. But its useful anyway.

**Reply by sfromis** [reply]
I'd suggest restraint on implementing active polling, as this may deplete the battery, possibly *much* faster.

**Reply by trichromic-ab** [reply]
Oh I fully agree on that. The "raison d'etre" of ZigBee is low power usage battery.
This is where we need learn to to select the right technology for the application needs.

**Reply by trichromic-ab** [reply]
> If you use `SensorRetain 1` and `SetOption89 1`, each sensor message will be published on its own sub-topic and the message will be retained. 

So you need to set both of these in order to have unique topics for each sensor and thus retain one message per sensor. Ok got that, thanks. However the topics are created with the device ID and not the friendly name when 'SetOption81 1'.
This is an issue because the device ID changes when a device is paired. So I rely on friendly name.
Wouldn't it be better if the SetOption89 provided friendly name topics when SetOption 81 is set?

**Reply by barbudor** [reply]
There's a SetOption for that too 🤣🤣🤣
SetOption112 (or ZbTopicName) will use the friendly name if defined insteald of the device ID

So you'll need all 3 of them 😂

**Reply by sfromis** [reply]
Also, SetOption100 (ZbNoPrefix) is nice, to avoid the somewhat tedious ZbReceived as an object level in the JSON format.

**Reply by trichromic-ab** [reply]
SetOption112 works like a charm. Thanks for that tip!

Being able to drop 'ZbReceived' with SetOption100 is a welcome surprise too. That will simplify things for me.
I should now be able to request an initial refresh using ZbInfo and parse the message in the same way as a ZbReceived message.

**Reply by trichromic-ab** [reply]
Sorry this went a bit off topic from the original query but the desired end result was the same and I have now achieved that quite well with your help.
I have created a script in HA that sends a ZbInfo to all bridges and I can trigger that on a button click to get the last cached sensor states (even if they are stale). So I'm happy now.
Thanks again. Hopefully this will guide others with similar needs.
